dtos/metric.go

```diff
@@ -0,0 +1,144 @@
+	Tags []MetricTag `json:"tags,omitempty"`
+	// Timestamp is the time and date the metric was collected.
+	Timestamp int64 `json:"timestamp" validate:"required"`
+}
+
+// MetricField defines a metric field associated with a metric
+type MetricField struct {
+	// Name is the identifier of the metric field
+	Name string `json:"name" validate:"edgex-dto-none-empty-string"`
+	// Value is measurement for the metric field
+	Value interface{} `json:"value" validate:"required"`
+}
+
+// MetricTag defines a metric tag associated with a metric
+type MetricTag struct {
+	// Name is the identifier of the metric tag
+	Name string `json:"name" validate:"edgex-dto-none-empty-string"`
+	// Value is tag vale for the metric tag
+	Value string `json:"value" validate:"required"`
+}
+
+// NewMetric creates a new metric for the specified data
+func NewMetric(name string, fields []MetricField, tags []MetricTag) (Metric, error) {
+	if err := ValidateMetricName(name, "metric"); err != nil {
+		return Metric{}, err
+	}
+
+	if len(fields) == 0 {
+		return Metric{}, errors.New("one or more metric fields are required")
+	}
+
+	for _, field := range fields {
+		if err := ValidateMetricName(field.Name, "field"); err != nil {
+			return Metric{}, err
+		}
+	}
+
+	if len(tags) > 0 {
+		for _, tag := range tags {
+			if err := ValidateMetricName(tag.Name, "tag"); err != nil {
+				return Metric{}, err
+			}
+		}
+	}
+
+	metric := Metric{
+		Versionable: common.NewVersionable(),
+		Name:        name,
+		Fields:      fields,
+		Timestamp:   time.Now().UnixNano(),
+		Tags:        tags,
+	}
+
+	return metric, nil
+}
+
+func ValidateMetricName(name string, nameType string) error {
+	if len(strings.TrimSpace(name)) == 0 {
+		return fmt.Errorf("%s name can not be empty or blank", nameType)
+	}
+
+	// TODO: Use regex to validate Name characters
+	return nil
+}
+
+// ToLineProtocol transforms the Metric to Line Protocol syntax which is most commonly used with InfluxDB
+// For more information on Line Protocol see: https://docs.influxdata.com/influxdb/v2.0/reference/syntax/line-protocol/
+// Line Protocol Syntax:
+//    <measurement>[,<tag_key>=<tag_value>[,<tag_key>=<tag_value>]] <field_key>=<field_value>[,<field_key>=<field_value>] [<timestamp>]
+// Examples:
+//    measurementName fieldKey="field string value" 1556813561098000000
+//    myMeasurement,tag1=value1,tag2=value2 fieldKey="fieldValue" 1556813561098000000
+//
+// Note that this is a simple helper function for those receiving this DTO that are pushing metrics to an endpoint
+// that receives LineProtocol such as InfluxDb or Telegraf
+func (m *Metric) ToLineProtocol() string {
+	isFirst := true
+	fields := ""
+	for _, field := range m.Fields {
+		// Fields section doesn't have a leading comma per syntax above so need to skip the comma on the first field
+		if isFirst {
+			isFirst = false
+		} else {
+			fields += ","
+		}
+		fields += fmt.Sprintf("%s=%s", field.Name, formatLineProtocolValue(field.Value))
```

Comment:
@lenny-intel for this kind of string concatenate logic, we should use string builder
https://pkg.go.dev/strings#Builder

> A Builder is used to efficiently build a string using Write methods. It minimizes memory copying. The zero value is ready to use.

Do you mind we open a new issue and refactor this part?

Reply:
@cloudxxx8 , good point!. No I do not mind. THX!

Reply:
https://github.com/edgexfoundry/go-mod-core-contracts/issues/728 opened for this